airframe-http/.jvm/src/main/scala/wvlet/airframe/http/router/HttpRequestMapper.scala

```diff
@@ -40,84 +47,186 @@ object HttpRequestMapper extends LogSupport {
+    def setValue(arg: MethodParameter, v: Option[Any]): Unit = {
+      methodArgs(arg.index) = v
+      // Use the method default argument value if exists
+        .orElse(arg.getMethodArgDefaultValue(controller))
+        // If no mapping is available, use the zero value
+        // TODO: Throw an error here when strict validation is enabled
+        .getOrElse(Zero.zeroOf(arg.surface))
+    }
+
+    def readContentBodyAsMsgPack: Option[MsgPack] = {
+      // Build the method argument instance from the content body for non GET requests
+      val contentBytes = adapter.contentBytesOf(request)
+
+      if (contentBytes.isEmpty) {
+        None
+      } else {
+        adapter.contentTypeOf(request).map(_.split(";")(0).toLowerCase()) match {
+          case Some("application/x-msgpack") =>
+            Some(contentBytes)
+          case Some("application/json") =>
+            // JSON -> msgpack
+            Some(MessagePack.fromJSON(contentBytes))
+          case Some("application/octet-stream") =>
+            // Do not read binary contents (e.g., uploaded file)
+            None
+          case _ =>
+            // Try parsing the content body as as JSON
+            Some {
+              Try(JSON.parse(contentBytes))
+                .map { jsonValue =>
+                  JSONCodec.toMsgPack(jsonValue)
+                }
+                .getOrElse {
+                  // If parsing as JSON fails, treat the content body as a regular string
+                  StringCodec.toMsgPack(adapter.contentStringOf(request))
                 }
             }
-            // If mapping fails, use the zero value
-            v.getOrElse(Zero.zeroOf(arg.surface))
         }
       }
+    }
+
+    // Populate the remaining function arguments
+    remainingArgs match {
+      case Nil =>
+      // Do nothing
+      case arg :: Nil =>
+        // For unary functions, we can omit the parameter name key in the request body
+        val argSurface = arg.surface
+        val argCodec   = codecFactory.of(argSurface)
+        readContentBodyAsMsgPack match {
+          case Some(msgpack) =>
+            // Read the content body as a MessagePack Map value
+            val v = MessagePack.newUnpacker(msgpack).unpackValue
+            val opt: Option[Any] = v match {
+              case m: MapValue if m.isEmpty =>
+                None
+              case m: MapValue =>
+                val mapValue = toCanonicalKeyNameMap(m)
+                // Extract the target parameter from the MapValue representation of the request content body
+                mapValue.get(CName.toCanonicalName(arg.name)) match {
+                  case Some(paramValue) =>
+                    // {"(param name)":(value)}
+                    Some(argCodec.fromMsgPack(paramValue.toMsgpack))
+                  case None =>
+                    // When the target parameter is not found in the MapValue, try mapping the content body as a whole
+                    argCodec.unpackMsgPack(msgpack)
```

Comment:
Though I know this is an original behavior of the parameter mapping in airframe-http, this behavior is sometimes convenient to take whole request body as a string, however, it might be a little dangerous because this happens unintentionally when we give a wrong parameter name.
```scala
def rpc1(name: String) = {}

{"name": "hello"} => "hello"
{"meme": "hello"} => """{"meme": "hello"}"""
```
If we need to do this intentionally, it would be better to use HttpRequest injection than relying on this fallback behavior.

Reply:
Yes. That's an issue and I'm worrying the compatibility of existing applications using airframe-http. For applications using sbt-airframe (e.g., Conductor), we can fix both of the client and server code at the same time, so we don't need to worry about the compatibility.

An idea is totally deprecating the whole content body mapping as you suggested, and ask users to use Request object.

Reply:
Filed this as https://github.com/wvlet/airframe/issues/1129. At least for RPC, which uses generated clients, we don't need to support the whole body mapping